### PR TITLE
Magnetometer & Sun Vector Attitude Filter Implementation

### LINKED ISSUE #190: Magnetometer and Sun Vector UKF Implementation
Per the title.

### PR DESCRIPTION
**Pull request by kylekrol**
# Magnetometer & Sun Vector Attitude Filter Implementation

Fixes #190.

### Summary of changes
- Added GRP attitude representation utility functions to convert to and from utility functions (deferring intensive unit testing to #193).
- Added double and float interfaces for all environment functions (this is helpful for switching between floats and doubles in the filter implementation itself).
- Updated the `q_eci_perifocal` constant to fix the sun vector model; appropriate unit tests were fixed as well.
- Initial attitude filter implementation when we have both magnetometer and sun sensor readings.
- Simple testcase for the attitude filter with magnetometer + sun sensor readings.

On the surface, the attitude filter seems to work. It's difficult to write good test cases using just PIO because numerical precision, at least so far, has been as issue. For example, the Frobenius norm of the update state covariance matrix is only accurate to two significant digits. Also, we currently only have a single testcase.

There are multiple tickets open to track the work on the attitude filter going forward: #187, #188, #189, #191, #193, and #197. This work includes:

* Developing more testcase and looking at long term filter behavior in MATLAB.
* Improving/verifying numerical stability specifically
* Implementing the magnetometer only Kalman gain update step.
* Adding GRP utility function testcases.

### Ptest Effects
NA

### Testing
See the added unit test cases and tickets listed above for further testing improvements to come.

### Constants
| Constant                 | Location             |
|--------------------------|----------------------|
| `gnc::constant::ukf_sigma_v`     | `src/gnc_attitude_estimator.cpp`   |
| `gnc::constant::ukf_sigma_u`     | `sr/gnc_attitude_estimator.cpp`   |
| `gnc::constant::ukf_sigma_b`     | `src/gnc_attitude_estimator.cpp`   |
| `gnc::constant::ukf_sigma_s`     | `src/gnc_attitude_estimator.cpp`   |
| `dt_thresh` (local scope)     | `src/gnc_attitude_estimator.cpp`   |
| `q_factor` (local scope)     | `src/gnc_attitude_estimator.cpp`   |
| `var_q` (local scope)     | `src/gnc_attitude_estimator.cpp`   |
| `var_g` (local scope)     | `src/gnc_attitude_estimator.cpp`   |
| `init_state` (local scope)     | `src/gnc_attitude_estimator.cpp`   |

### Documentation Evidence
See inline documentation and the paper mentioned in the file documentation in `include/gnc/attitude_estimator.hpp`.

### CODE REVIEW

File: src/gnc_environment.cpp [hunk 1]
```diff
@@ -68,13 +68,25 @@ void sun_vector(double t, lin::Vector3f &s) {
...
+void magnetic_field(double t, lin::Vector3d const &r, lin::Vector3f &b) {
   geomag::Vector in, out;
   in.x = r(0);
   in.y = r(1);
   in.z = r(2);
   out = geomag::GeoMag(constant::init_dec_year + t / (365.0 * 24.0 * 60.0 * 60.0),in, geomag::WMM2020);
   b = { out.x, out.y, out.z };
 }
+
+void magnetic_field(double t, lin::Vector3d const &r, lin::Vector3d &b) {
```
Comment (tanishqaggarwal) [inline src/gnc_environment.cpp:86]: Won't this function cause an infinite recursion?

File: include/gnc/constants.hpp [hunk 2]
```diff
@@ -60,7 +60,7 @@ GNC_TRACKED_CONSTANT(constexpr static double, earth_perihelion_time, -6287982.0)
 
 GNC_TRACKED_CONSTANT(constexpr static double, earth_period, 365.256363004 * 24.0 * 60.0 * 60.0);
 
-GNC_TRACKED_CONSTANT(constexpr static lin::Vector4d, q_eci_perifocal, 0.127456632677880, -0.158124280715206, 0.762378784011859, -0.614434787689722);
+GNC_TRACKED_CONSTANT(constexpr static lin::Vector4d, q_eci_perifocal, 0.127456632677880, -0.158124280715206, 0.762378784011859, 0.614434787689722);
```
Comment (nhz2) [inline include/gnc/constants.hpp]: Why is this changed and does it need to be changed in MATLAB?

File: test/test_all/environment_test.cpp [hunk 3]
```diff
@@ -60,23 +60,23 @@ void test_environment_sun_vector() {
   // Calculate and check for t=0.0 (comparing against MATLAB)
   gnc::env::sun_vector(0.0L, s);
   TEST_ASSERT_FLOAT_VEC_NEAR(1e-5, s, lin::Vector3f({
-     0.997341623364311f,
```
Comment (nhz2) [inline test/test_all/environment_test.cpp]: Why did this change?

File: include/gnc/utilities.hpp [hunk 4]
```diff
@@ -47,6 +47,27 @@ template <typename T>
 constexpr void quat_cross_mult(lin::Vector<T, 4> const &q1,
         lin::Vector<T, 4> const &q2, lin::Vector<T, 4> &res);
 
+/** @fn quat_to_grp
+ *  @param[in]  q Input quaternion.
+ *  @param[in]  a 
+ *  @param[in]  f
+ *  @param[out] p Generalized rodriquez parameter representation.
```
Comment (nhz2) [inline include/gnc/utilities.hpp]:
```suggestion
 *  @param[out] p Generalized Rodrigues parameter representation.
```

File: include/gnc/utilities.hpp [hunk 5]
```diff
@@ -47,6 +47,27 @@ template <typename T>
...
+ *  Converts a quaternion attitude representation to generalized rodriquez
```
Comment (nhz2) [inline include/gnc/utilities.hpp]:
```suggestion
 *  Converts a quaternion attitude representation to generalized Rodrigues
```

File: include/gnc/utilities.hpp [hunk 6]
```diff
@@ -47,6 +47,27 @@ template <typename T>
...
+ *  parameters. There is no explicit handling of NaNs built into this function;
+ *  however, a finite input will always yield a finite result.
+ *  Source: "Unscented Filtering for Spacecraft Attitude Estimation"
+ *          by Markley and Crasidis */
+template <typename T>
+constexpr void quat_to_qrp(lin::Vector<T, 4> const &q, T a, T f, lin::Vector<T, 3> &p);
+
+/** @fn grp_to_quat
+ *  @param[in]  p Input rogriquez parameters.
```
Comment (nhz2) [inline include/gnc/utilities.hpp]:
```suggestion
 *  @param[in]  p Input Rodrigues parameters.
```

File: include/gnc/attitude_estimator.hpp [hunk 7]
```diff
@@ -144,8 +181,20 @@ struct AttitudeEstimate {
   bool is_valid;
 };
 
-/** @fn attitude_estimator_reset
- *  Initializes the attitude filter state using the triad method.
+/** @brief Initializes the attitude filter state.
+ *
+ *  @param[out] state      Attitude estimator state to be reset/initialized.
+ *  @param[in]  t          Time since the PAN epoch (units of seconds).
+ *  @param[in]  q_body_eci Quaternion rotating from the body frame to ECI.
```
Comment (nhz2) [inline include/gnc/attitude_estimator.hpp]:
```suggestion
 *  @param[in]  q_body_eci Quaternion rotating from ECI to the body frame.
```

File: src/gnc_attitude_estimator.cpp [hunk 8]
```diff
@@ -55,26 +387,92 @@ AttitudeEstimate::AttitudeEstimate()
...
-  GNC_ASSERT_NORMALIZED(q_eci_body);
+    double t, lin::Vector3d const &r_ecef, lin::Vector3f const &b_body,
+    lin::Vector3f const &s_body) {
+  GNC_ASSERT_NORMALIZED(s_body);
 
   // TODO : Implement this
-  state = AttitudeEstimatorState();
+  // https://github.com/pathfinder-for-autonomous-navigation/psim/issues/187
+  lin::Vector4f q_eci_body = lin::nans<float, 4, 1>();
+  attitude_estimator_reset(state, t, q_eci_body);
```
Comment (nhz2) [inline src/gnc_attitude_estimator.cpp]:
```suggestion
  lin::Vector4f q_body_eci = lin::nans<float, 4, 1>();
  attitude_estimator_reset(state, t, q_body_eci);
```

File: src/gnc_attitude_estimator.cpp [hunk 9]
```diff
@@ -55,26 +387,92 @@ AttitudeEstimate::AttitudeEstimate()
...
-  state = AttitudeEstimatorState();
-  estimate = AttitudeEstimate();
+  // Ensure we have a valid state and all the required inputs
+  if (!state.is_valid ||
+      !lin::all(lin::isfinite(data.r_ecef)) ||
+      !lin::all(lin::isfinite(data.b_body)) ||
+      !lin::all(lin::isfinite(data.w_body)) ||
+      !lin::isfinite(data.t)) {
+    // Set the state and estimate to be invalid
+    state = AttitudeEstimatorState();
```
Comment (nhz2) [inline src/gnc_attitude_estimator.cpp:442]: Can the filter just propagate without sensor readings? Or does the state need to be reset.

File: src/gnc_attitude_estimator.cpp [hunk 10]
```diff
@@ -28,16 +28,348 @@
...
+    {
+      UkfMatrix5x5 Q, R;
+      lin::qr(state.P_vv, Q, R);
+      lin::backward_sub(R, Q, lin::transpose(Q).eval()); // Q = inv(P_yy)    
+      K = state.P_xy * Q;
+    }
+
+    // Calculate this steps measurement
+    UkfVector5 z_new {
+      lin::atan(data.s_body(1) / data.s_body(0)),
```
Comment (nhz2) [inline src/gnc_attitude_estimator.cpp:356]: This can be NAN if `s_body(0)==0`

File: include/gnc/constants.hpp (hunk 2 above)
Comment (kylekrol) [reply include/gnc/constants.hpp]: So at some point this got changed in MATLAB but wasn't updated in the C++. As a result, our sun vector environment function was just wrong. Stewart and I figured this out while debugging the filter.

File: test/test_all/environment_test.cpp (hunk 3 above)
Comment (kylekrol) [reply test/test_all/environment_test.cpp]: Same reason as above. The sun vector model got updated in MATLAB at this point and not in C++.

File: src/gnc_attitude_estimator.cpp (hunk 9 above)
Comment (kylekrol) [reply src/gnc_attitude_estimator.cpp:442]: You can't just propagate without sensor readings, at least that's how it's coded now. In terms of our use case however, there won't be a cc where we don't have at least magnetometer and gyroscope data. If for some reason we don't have that information, we have bigger problems than our attitude filter too lol

File: src/gnc_attitude_estimator.cpp (hunk 10 above)
Comment (kylekrol) [reply src/gnc_attitude_estimator.cpp:356]: This is a little concerning... However, in cxx a finite float over zero results in plus or minus infinity (IEEE floats at least) and according to the documentation the `std::atan` function (which is what `lin::atan` essentially calls) returns `pi/2` and `-pi/2` for `inf` and `-inf` respectively. In theory we should be alright.
This code will get replaced when the new sun vector variance handling scheme is incorporated (issue for that coming soon).

Links:
https://en.cppreference.com/w/cpp/numeric/math/atan
https://stackoverflow.com/questions/4745311/c-division-by-0

File: test/test_all/environment_test.cpp (hunk 3 above)
Comment (nhz2) [reply test/test_all/environment_test.cpp]: In this branch in MATLAB I'm getting:
```
>> config
>> env_sun_vector(0)

ans =

    0.9973
   -0.0668
   -0.0290
```

Comment (nhz2) [reply test/test_all/environment_test.cpp]: This makes sense because PAN epoch is in march, near the vernal equinox.

File: src/gnc_environment.cpp [hunk 11]
```diff
@@ -68,13 +68,25 @@ void sun_vector(double t, lin::Vector3f &s) {
   utl::rotate_frame(q_eci_perifocal_f, s);
 }
 
-void magnetic_field(double t, lin::Vector3f const &r, lin::Vector3f &b) {
+void sun_vector(double t, lin::Vector3d &s) {
```
Comment (tanishqaggarwal) [inline src/gnc_environment.cpp:71]: Same with this one

**Review by tanishqaggarwal — APPROVED** [review]
Just a couple of comments

File: src/gnc_environment.cpp (hunk 11 above)
Comment (kylekrol) [reply src/gnc_environment.cpp:71]: No, because of the float `lin::Vector3f` overload just above it.

File: src/gnc_environment.cpp (hunk 1 above)
Comment (kylekrol) [reply src/gnc_environment.cpp:86]: Same with this one; there is the float vector overload.

File: test/test_all/environment_test.cpp (hunk 3 above)
Comment (kylekrol) [reply test/test_all/environment_test.cpp]: Fixed.

File: include/gnc/constants.hpp (hunk 2 above)
Comment (kylekrol) [reply include/gnc/constants.hpp]: Reverted these changes and fixed the attitude filter testcase accordingly.

**Review by nhz2 — APPROVED** [review]
